File: OrchidCore/src/main/java/com/eden/orchid/utilities/ClogSpells.java
package com.eden.orchid.utilities;

import com.caseyjbrooks.clog.parseltongue.Spell;
import com.eden.orchid.Orchid;

import java.net.MalformedURLException;
import java.net.URL;

public class ClogSpells {

    @Spell
    public static String baseUrl(Object object) {
        return Orchid.getInstance().getContext().getSite().getBaseUrl();
    }

    @Spell
    public static String baseUrlScheme(Object object) throws MalformedURLException {
        return new URL(ClogSpells.baseUrl(object)).getProtocol();
    }

    @Spell
    public static String baseUrlHost(Object object) throws MalformedURLException {
        return new URL(ClogSpells.baseUrl(object)).getHost();
    }

    @Spell
    public static int baseUrlPort(Object object) throws MalformedURLException {
        return new URL(ClogSpells.baseUrl(object)).getPort();
    }

    @Spell
    public static String baseUrlRoot(Object object) throws MalformedURLException {
        URL url = new URL(ClogSpells.baseUrl(object));

        String urlRoot = "";
        urlRoot += url.getProtocol() + "://";
        urlRoot += url.getHost();

        if(url.getPort() != -1 && url.getPort() != 80) {
            urlRoot += ":" + url.getPort();
        }

        return urlRoot;
    }

    @Spell
    public static String orchidVersion(Object object) {
        return Orchid.getInstance().getContext().getSite().getOrchidVersion();
    }

    @Spell
    public static String env(Object object) {
        return Orchid.getInstance().getContext().getSite().getEnvironment();
    }

    @Spell
    public static String version(Object object) {
        return Orchid.getInstance().getContext().getSite().getVersion();
    }

}

File: OrchidCore/src/main/kotlin/com/eden/orchid/impl/compilers/clog/ClogSpells.kt
package com.eden.orchid.impl.compilers.clog

import com.caseyjbrooks.clog.parseltongue.Spell
import com.eden.orchid.api.OrchidContext
import java.net.MalformedURLException
import java.net.URL
import javax.inject.Inject
import javax.inject.Provider

@Suppress("UNUSED_PARAMETER")
class ClogSpells
@Inject
constructor(
    val contextProvider: Provider<OrchidContext>
) {

    @Spell
    fun baseUrl(value: Any): String {
        return contextProvider.get().site.baseUrl
    }

    @Spell
    @Throws(MalformedURLException::class)
    fun baseUrlScheme(value: Any): String {
        return URL(baseUrl(value)).protocol
    }

    @Spell
    @Throws(MalformedURLException::class)
    fun baseUrlHost(value: Any): String {
        return URL(baseUrl(value)).host
    }

    @Spell
    @Throws(MalformedURLException::class)
    fun baseUrlPort(value: Any): Int {
        return URL(baseUrl(value)).port
    }

    @Spell
    @Throws(MalformedURLException::class)
    fun baseUrlRoot(value: Any): String {
        val url = URL(baseUrl(value))

        var urlRoot = ""
        urlRoot += url.protocol + "://"
        urlRoot += url.host

        if (url.port != -1 && url.port != 80) {
            urlRoot += ":" + url.port
        }

        return urlRoot
    }

    @Spell
    fun orchidVersion(value: Any): String {
        return contextProvider.get().site.orchidVersion
    }

    @Spell
    fun env(value: Any): String {
        return contextProvider.get().site.environment
    }

    @Spell
    fun version(value: Any): String {
        return contextProvider.get().site.version
    }

}
